feat: priority donation 구현

hdh853 opened the pull request:
## 🔗 관련 이슈
<!-- 이슈 번호를 연결해주세요 -->
closes #16 

## 📌 작업 내용
### 구현 사항
<!-- 구현내용에 대해서 설명해주세요. -->
<!-- 예시 :
- 이슈 템플릿, PR 템플릿 작성
- READEME 파일 최초 작성
 --> 
- thread 구조체에 기부받은 우선순위, 잡고있는 락 목록, 대기하고있는 락 멤버 추가
- 기존에 사용되던 priority를 donation_priority로 교체
- 락 취득시 사용되는 우선순위 기부 함수 추가
- 락 반납시 사용되는 우선순위 반환 함수 추가

### 변경된 파일
<!-- 구현 내용으로 인해 변경된 파일에 대해 설명해주세요 -->
<!-- 예시:
- README.md
  - readme 파일 최초 생성
- issue_template.md
  - 이슈 템플릿 최초 생성
-->
- thread.h
  - donation_priority, lock_holding, lock_waiting 멤버  추가
- thread.c
  - thread_create, init_thread, thread_set_priority, thread_get_priority, priority_greater 에서 사용되던 priority를 donation_priority 로 교체
- synch.c
  - lock_acquire시 취득하려는 락이 우선순위가 낮은 쓰레드에게 점유되어 있다면, 우선순위 기부하는 로직 추가
  - 우선순위 기부 함수(donation_priority) 추가
  - lock_release시 잡고있는 락 순회하며 해당 락들의 대기자들에게 우선순위 기부받는 로직 추가
  - 우선순위 반환 함수(retrieve_priority) 추가

### 추가 사항
<!-- 트러블 슈팅 등 추가로 설명할 내용이 있으시다면 아래에 작성해주세요. -->

## ✅ 셀프 체크리스트
<!-- 체크리스트를 모두 완료한 후 제출해주세요 -->
<!-- 완료 처리를 표시하려면 [x], 완료되지 않았다면 [ ] -->
- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 주석이나 디버깅 코드를 제거했나요?
- [x] 관련 테스트가 모두 통과하나요?
- [x] 커밋 메시지가 컨벤션에 맞나요?

hdh853 committed:
style: synch.h 포멧팅 적용

hdh853 committed:
feat: priority donation 구현
- 우선순위 검사 대상을 기부받은 우선순위로 변경
- thread 구조체에 기부받은 우선순위, 잡고있는 락 리스트, 대기하고있는 락 구조체 멤버로 추가
- synch.c에서 락 취득시 우선순위 기부, 락 반납시 우선순위 반환 구현

hdh853 committed:
chore: retrieve_priority 함수 주석 추가

hyunmuam commented on threads/synch.c:
```diff
@@ -223,6 +265,35 @@ bool lock_try_acquire(struct lock* lock)
+            }
+        }
+    }
+}
```
여기도 중첩된 if문을 early return 패턴으로 개선하면 가독성이 올라갈 것 같습니다.
예: if (조건) return; 형태로 예외 케이스를 먼저 처리

현재 waiter리스트를 정렬해서 `list_begin`을 해주는데, `sema_*`, `conv_*`에서 했던 것 처럼 list_max를 사용하는건 어떠신가요?

hdh853 replied:
그렇네요. sort 후 begin 보다 list_min으로 조회하는게 코드도 깔끔하고 시간복잡도도 나을것 같네요.
바로 반영하겠습니다 감사합니다.

hyunmuam commented on threads/synch.c:
```diff
@@ -163,6 +163,35 @@ static void sema_test_helper(void* sema_)
+        }
+    }
+}
+
```
이중 if문을 각 조건을 독립적인 if문으로 분리하면 
코드 흐름을 파악하기 더 쉬울 것 같습니다.

hyunmuam commented on threads/thread.c:
```diff
@@ -316,17 +316,24 @@ void thread_set_priority(int new_priority)
+        if (new_priority < target->donation_priority) {
+            thread_yield();
+        }
     }
```
여기도 이중 if문을 각 조건을 독립적인 if문으로 분리하면
코드 흐름을 파악하기 더 쉬울 것 같습니다.

hyunmuam reviewed (commented):
구현 로직은 제 방식과 달라서 세부 로직 리뷰보다는, 코드 구조 측면에서 주로 리뷰를 남겼습니다.

현재 조건문이 엄청나게 긴데, 중첩까지 되어 있어 각 분기의 조건을 파악하기 어려웠습니다. early return 패턴으로 개선하시면 예외 조건을 먼저 처리하고 주요 로직에 집중할 수 있어 가독성이 크게 향상될 것 같습니다.
참고: 조건 반전 -> 즉시 return 순서로 리팩토링하면 들여쓰기 depth를 줄일 수 있습니다.

+) 연결하신 이슈가 #17인데, 제 PR 번호인 것 같습니다. 다시 한 번 확인해주시고 이슈 연결 해주세요. :)

hdh853 committed:
refactor: 피드백 반영
- if 중첩문 해결 위해 early return 패턴 적용
- retrieve_priority에서 waiter목록 조회할 때 (기존)sort 후 list_begin 에서 (수정)list_min 으로 조회하도록 변경

hdh853 commented:
if문을 early return으로 쪼개는 방식으로 수정해봤는데, 확실히 이전보다 조건 검사 의도가 더 잘 나타나는것 같고, 코드도 깔끔한것 같네요. 사용하는 습관 들여보겠습니다!
피드백 감사합니다.

hyunmuam reviewed (commented):
굳~
